### DISCUSSION TITLE
what do you think about xtask? will it replace cargo-make?

### DISCUSSION THREAD

**Discussion by dzmitry-lahoda** [sagiegurari/cargo-make#717] [General]
xtask is part of main repo and based on aliases (which are part of cargo):

https://github.com/rust-lang/rust-analyzer/blob/master/xtask/src/main.rs

if .cargo/config supports workspace sharing of aliases (if not, there is some work for sharing stuff in root toml).

then seems solution is very good and seems way ecosystem will go.

other things i see that people stick to usual old school make or wrap into nix (nix by default uses make). so having usual make makes you nix compatible too.

**Comment by sagiegurari** [comment]
I think xtasks are pretty cool. 
will it replace? my guess probably not. same as cargo-make won't replace make and vice versa. there are many tools. each has pros/cons and they take very different approaches.

xtasks and cargo-make take a very different path. 
for example, by using xtasks i would ask:
* need to understand the difference between definition and code. would you write a complex build script in rust? for example looping over files and doing some action on them? or installing dependencies like wasm, audit, outdated, .... (what happens if they are already installed, or old version? cargo-make does it for you automatically). what about tasks depending on other tasks? what about conditions if to run or not? at the end, you will have to code everything instead of defining it. its a lot more work. in your example, they have a lot of rust files to define and handle those tasks. its really a lot of work.
* you will need to start downloading a lot of dependencies in your xtasks to have basic capabilities in the xtasks (per project!)
* xtasks forces your project to be a workspace. personally I try to avoid the extra complexity (not everyone will see it as complex, but i like a single simple crate if possible).
* even the example you gave, they jump from rust into shell ```cmd!(sh, "cargo install cargo-fuzz").run()?;``` so i'll ask, is that easy and clear (readable?). it ends up a mesh of things.
* I saw projects using cargo-make for non rust projects. i personally use them for few c projects, but i saw others using it for stuff like python. its just a task runner after-all.

The biggest pro i see in xtasks, its bundled with the project. cargo-make only bundles the makefile.toml with it, but the binary is installed separately. 

again, all of the above is my personal view. i'm SUPER bias since i wrote cargo-make based on my vision of a task runner after working with many tasks runners in other platforms like maven, npm, grunt, gulp, shell files, makefiles, cmake and so on...
i'm not saying its perfect or even close, but I like the 'definition' approach more than coding a build script in rust.

**Comment by dzmitry-lahoda** [comment]
just checked - aliases work workspace somewhat. not tested xtask

<img width="587" alt="image" src="https://user-images.githubusercontent.com/757125/192487539-e55fb1e9-d916-4235-aa84-eae15cba0a6a.png">

but yeah - kind of will need copy paste some for sure.

**Reply by sagiegurari** [reply]
I don't think aliases are the issue. i think both tools take very different approach which is why i don't see it replaced. which was your original question.
as you can see from github only search (there are many private projects using it that are not open source):
https://github.com/search?utf8=%E2%9C%93&q=filename%3AMakefile.toml&type=Code&s=indexed
there are many different types of makefiles and how people do things. not all are easily converted to rust code.
a build script is more than just firing one command. its a flow.
thats why for example, my github actions are really simple, they just call cargo make ci-flow. no need for a ton of yaml steps in each project. and its much simpler to run same flow locally to see everything is ok.
building flows in shell/xtasks/... might be more difficult. again, task runner does a bit more than just run a single task.
